dom0/runner.py

```diff
@@ -75,6 +75,21 @@ def __init__(self):
             ]
         )
 
+    def shutdown_sd_vms(self):
+        """
+        Shut down sd-workstation-tagged VMs.
+
+        This should be done before dom0 tests, to ensure that the
+        AppVMs tested have the latest TemplateVM changes.
+        """
+        self.logging.info("Shutting down SecureDrop Workstation VMs")
+        sdw_vms = [vm for vm in self.q.domains if "sd-workstation" in vm.tags]
+
+        for vm in sdw_vms:
```

Comment:
Just noting that qvm-shutdown (which is Python importable, see https://github.com/QubesOS/qubes-core-admin-client/blob/main/qubesadmin/tools/qvm_shutdown.py) has a graceful loop that shuts the VMs down in order so you can avoid using --force. Don't think there's anything necessarily wrong with this approach though.

Reply:
My understanding is that `--force` in this case isn't a forced immediate halt like qvm-kill but just bypasses VM dependencies and issues a regular ACPI shutdown signal (so you can shut down a NetVM that has qubes using it, for example).